### ISSUE TITLE
NativeScript Interop Compatibility with raw C code

### ISSUE THREAD

**Issue by NathanaelA** [NativeScript/ios-runtime#816]
Ok, I have a call that needs basically an array of null terminated strings; however, the array of pointers need to be from the same string.   Normally I would just make a ObjC function; but I'm curious to know if NativeScript can actually do this without me having to create a ObjC function, to handle this...

So, say I have this string:  ` "Hi\0This\0test\0";`  My array needs to have a pointers to 0, 3, 8.

Is there a way to do this in NativeScript, i.e. 
```
var buffer=interop.alloc(13);
// copy js string into buffer 
// create an array of pointers to spot[0], spot[3], spot[8];
```

My thought is something like this:
```
var str= "Hi\0This\0test\0";
var buffer = malloc(13 * interop.sizeof(interop.types.unichar));

// Copy string over manually  (Might instead use NSString.getCharacters)
var reference = new new interop.Reference(interop.types.unichar, buffer);
for (var i=0;i<13;i++) reference[i] = str[i];

// Create the 3 pointer array
var bagvals = interop.alloc(3 * interop.sizeof(interop.Pointer));
var bagvalsRef = new interop.Reference(interop.Pointer, bagvals);
bagvalsRef[0] = interop.handleof(buffer.add(0));
bagvalsRef[1] = interop.handleof(buffer.add(3));
bagvalsRef[2] = interop.handleof(buffer.add(5));
// Do whatever with this object

// Finally free it
free(buffer);
```
Am I correct that bagvals would be an array of pointers to those three memory locations; or am I missing something in my understanding of the interop stuff.

**Comment by tdermendjiev** [comment]
Hello @NathanaelA ,
If I understand correctly, you are trying to manipulate *bagvals* as a C pointer. The value returned from *interop.alloc* method is an instance of *interop.Pointer* class which points to the first element of the allocated memory and as such (according to the docs(https://docs.nativescript.org/runtimes/ios/types/c-pointers)) it has no subscript members, i.e. it cannot be manipulated as an array in your javascript code. If you need subscript access you can only use *interop.Reference* class.
If passed to a native function though *bagvals* is a perfectly valid C array and it comes to the proper parameter's type passed to it. I used the following function to reproduce this scenario:

```
void logNativeStrings(uint16_t** strings, int count) {
    for (int i = 0; i < count; i++) {
        std::u16string str(reinterpret_cast<char16_t*>(strings[i]));
        NSString *string = [[NSString alloc] initWithCharacters:(strings[i]) length:str.length()];
        NSLog(@"%zu  <-----", str.length());    
        NSLog(@"%@", string);
    }
}
```

**Note**
*I used `uint16_t` instead of `wchar_t` as the string's underlying type in JSC is 16 bits wide while in iOS `wchar_t `is 32 bits.* 

I made a couple of changes in your javascript code in order to fix the pointer arithmetic and this is the final version of the code which I tested with:

```
 const str= "Hi\0This\0tests\0";
 const len=str.length;
 const charSize = interop.sizeof(interop.types.unichar);
 const buffer = malloc(len * charSize);
                                                          
 // Copy string over manually  (Might instead use NSString.getCharacters)
  var reference = new interop.Reference(interop.types.unichar, buffer);
 for (var i=0;i<len;i++) {
     reference[i] = str[i];
  }
 
  // Create the 3 pointer array
  var bagvals = interop.alloc(3 * interop.sizeof(interop.Pointer));
  var bagvalsRef = new interop.Reference(interop.Pointer, bagvals);
  bagvalsRef[0] = interop.handleof(buffer.add(charSize*0));
  bagvalsRef[1] = interop.handleof(buffer.add(charSize*3));
  bag
  logNativeStrings(bagvals, 3);
  
  // Finally free it
  free(buffer);
```
So basically you can prepare your arrays in such way in javascript. Any additional context of how you intent to use the array would be useful for further discussions on this matter.

**Comment by NathanaelA** [comment]
Thanks, so I was on the correct path!   I assume the bagValRef[2] line should look like the [1] line...
